src/mono/wasm/Wasm.Build.Tests/IcuTestsBase.cs

```diff
@@ -102,28 +106,105 @@ protected string GetProgramText(string testedLocales, bool onlyPredefinedCulture
+    protected async Task TestIcuShards(string config, string templateType, bool aot, string shardName, string testedLocales, GlobalizationMode globalizationMode, bool onlyPredefinedCultures=false)
     {
-        string projectName = $"shard_{Path.GetFileName(shardName)}_{buildArgs.Config}_{buildArgs.AOT}";
-        bool dotnetWasmFromRuntimePack = !(buildArgs.AOT || buildArgs.Config == "Release");
-
-        buildArgs = buildArgs with { ProjectName = projectName };
+        bool isBrowser = templateType == "wasmbrowser";
+        string icuProperty = isBrowser ? "BlazorIcuDataFileName" : "WasmIcuDataFileName"; // https://github.com/dotnet/runtime/issues/94133
         // by default, we remove resource strings from an app. ICU tests are checking exception messages contents -> resource string keys are not enough
-        string extraProperties = $"<WasmIcuDataFileName>{shardName}</WasmIcuDataFileName><UseSystemResourceKeys>false</UseSystemResourceKeys>";
+        string extraProperties = $"<{icuProperty}>{shardName}</{icuProperty}><UseSystemResourceKeys>false</UseSystemResourceKeys><RunAOTCompilation>{aot}</RunAOTCompilation>";
         if (onlyPredefinedCultures)
             extraProperties = $"{extraProperties}<PredefinedCulturesOnly>true</PredefinedCulturesOnly>";
-        buildArgs = ExpandBuildArgs(buildArgs, extraProperties: extraProperties);
+        await BuildAndRunIcuTest(config, templateType, aot, testedLocales, globalizationMode, extraProperties, onlyPredefinedCultures, icuFileName: shardName);
+    }
 
+    protected (BuildArgs buildArgs, string projectFile) CreateIcuProject(
+        string config,
+        string templateType,
+        bool aot,
+        string testedLocales,
+        string extraProperties = "",
+        bool onlyPredefinedCultures=false)
+    {
+        string id = $"icu_{config}_{aot}_{GetRandomId()}";
+        string projectFile = CreateWasmTemplateProject(id, templateType);
+        string projectDirectory = Path.GetDirectoryName(projectFile)!;
+        string projectName = Path.GetFileNameWithoutExtension(projectFile);
+        var buildArgs = new BuildArgs(projectName, config, aot, id, null);
+        buildArgs = ExpandBuildArgs(buildArgs);
+        AddItemsPropertiesToProject(projectFile, extraProperties: extraProperties);
+        
+        string programPath = Path.Combine(projectDirectory, "Program.cs");
         string programText = GetProgramText(testedLocales, onlyPredefinedCultures);
+        File.WriteAllText(programPath, programText);
         _testOutput.WriteLine($"----- Program: -----{Environment.NewLine}{programText}{Environment.NewLine}-------");
-        (_, string output) = BuildProject(buildArgs,
-                        id: id,
+
+        bool isBrowser = templateType == "wasmbrowser";
+        string mainPath = isBrowser ? Path.Combine("wwwroot", "main.js") : "main.mjs";
+        var replacements = isBrowser ? new Dictionary<string, string> { 
```

Comment:
Don't we need to exit on console as well?

Reply:
Exit is already there, so no replace is needed. If I used the replace pairs from `replacements`, we would get:
originally: `runMainAndExit()`
after: `runMainAndExitAndExit()`
We have 2 options:
1) phrase 2 replacement pairs instead of only one:
`{ "runMain()", "runMainAndExit()" }` and `{ ", runMain }", ", runMainAndExit }" }`
2) do not touch console app - that is the current one.
Do you think we should add a comment about it? Or change the approach?

Reply:
It's fine this way, thank you!